outline/proxy/config.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This package provides a highly abstracted client interface for use by
+// the platform code via `gobind`.  The exposed functions allow the platform
+// code to convert an opaque proxy configuration string into an opaque client
+// object and check that the client object is functioning (i.e. able to connect).
+// This package currently only offers basic Shadowsocks support, but over time it
+// may grow to support other protocols without altering the API as seen by the
+// platform code.
+package proxy
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	shadowsocks "github.com/Jigsaw-Code/outline-ss-server/client"
+	"github.com/eycorsican/go-tun2socks/common/log"
+)
+
+// Must match the ShadowsocksSessionConfig interface in Typescript.
+type configJSON struct {
+	Host     string
+	Port     uint16
+	Password string
+	Method   string
+	Prefix   string
+}
+
+// Client provides a transparent container for [client.Client] that
+// is exportable (as an opaque object) via gobind.
+type Client struct {
```

Comment:
I think this Client really only needs the `DialTCP` and `ListenUDP` methods. Perhaps we just expose those here?

Reply:
Also, maybe this should live on a separate client.go file so it's not conflated with the config stuff.

Reply:
TODO:
- Create a separate proxy/client.go file with the `Client` definition. The definition must stand alone and not depend on `shadowsocks.Client`. It should be the other way around.
- Create a `proxy/config/` for the config-based proxy creation.
- Have the shadowsocks config logic in `proxy/config/shadowsocks.go`, so we can put other protocols there too. It's important to separate the config stuff because there may be multiple ways to create the Client objects.

Perhaps we need a design discussion though. I really want to separate TCP from UDP. Ultimately we need to support the tun2socks handlers and direct dialers for creating sockets and the connectivity test.